src/main/kotlin/Main.kt

```diff
@@ -4,35 +4,56 @@ import com.google.protobuf.ByteString
 import franz.ProducerBuilder
 import franz.WorkerBuilder
 import kotlinx.coroutines.experimental.runBlocking
+import mu.KotlinLogging
 import se.zensum.idempotenceconnector.IdempotenceStore
-import java.util.Random
+import se.zensum.scheduler_proto.Scheduler.Task
+import java.nio.charset.Charset
+import java.time.Instant
+import java.util.*
+import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.DelayQueue
 import java.util.concurrent.Delayed
 import java.util.concurrent.TimeUnit
 import java.util.concurrent.atomic.AtomicBoolean
-import se.zensum.scheduler_proto.Scheduler.Task
-import java.nio.charset.Charset
-import mu.KotlinLogging
+
 const val TOPIC = "scheduler"
 
+private val logger = KotlinLogging.logger("scheduler")
+
 // A scheduled task
-data class ScheduledTask(private val task: Task): Delayed {
+data class ScheduledTask(private val task: Task, private val time: Long): Delayed {
+    constructor(task: Task): this(task, task.time) {}
+
     override fun compareTo(other: Delayed?): Int =
         getDelay(TimeUnit.SECONDS)
             .compareTo(other!!.getDelay(TimeUnit.SECONDS))
-    override fun getDelay(unit: TimeUnit?): Long =
-        unit!!.convert(fromNow(task.time), TimeUnit.SECONDS)
 
+    override fun getDelay(unit: TimeUnit?): Long =
+        unit!!.convert(fromNow(time), TimeUnit.SECONDS)
+    fun isExpired(): Boolean = task.expires != 0L && epoch() > task.expires
+    fun isRepeating() = task.interval != 0 && task.expires != 0L
+    fun isRemoved() = task.remove
+    private fun shouldReschedule() = isRepeating() && !isExpired()
```

Comment:
Might lead to one rescheduling after the task is expired since the "isExpired" refers to present time rather than when it will be run next time.